core/src/main/java/kafka/server/share/SharePartition.java

```diff
@@ -1791,10 +1875,23 @@ be removed once all the messages (0-99) are acknowledged (ACCEPT or REJECT).
             long firstKeyToRemove = cachedState.firstKey();
             long lastKeyToRemove;
             NavigableMap.Entry<Long, InFlightBatch> entry = cachedState.floorEntry(lastOffsetAcknowledged);
+            // If the lastOffsetAcknowledged is equal to the last offset of entry, then the entire batch can potentially be removed.
             if (lastOffsetAcknowledged == entry.getValue().lastOffset()) {
-                startOffset = cachedState.higherKey(lastOffsetAcknowledged);
+                if (initialReadGapOffset != null) {
+                    // This case will arise if we have a situation where there is an acquirable gap after the lastOffsetAcknowledged.
+                    // Ex, the cachedState has following state batches -> {(0, 10), (11, 20), (31,40)} and al these batches are acked.
+                    // In this case, lastOffsetAcknowledged will be 20, but we cannot simply move the start offset to the first offset
+                    // of next cachedState batch. The startOffset should be at 21, because we have an acquirable gap there.
+                    startOffset = Math.min(initialReadGapOffset.gapStartOffset(), cachedState.higherKey(lastOffsetAcknowledged));
+                } else {
+                    // If initialReadGapOffset is null, that means the cachedState does not have any acquirable gaps.
+                    // We can simply move the start offset to the first offset of the next cachedState batch.
+                    startOffset = cachedState.higherKey(lastOffsetAcknowledged);
+                }
```

Comment:
The code can be simpler though:

```
startOffset = cachedState.higherKey(lastOffsetAcknowledged);
if (isInitialReadGapOffsetWindowActive()) {
                    // This case will arise if we have a situation where there is an acquirable gap after the lastOffsetAcknowledged.
                    // Ex, the cachedState has following state batches -> {(0, 10), (11, 20), (31,40)} and al these batches are acked.
                    // In this case, lastOffsetAcknowledged will be 20, but we cannot simply move the start offset to the first offset
                    // of next cachedState batch. The startOffset should be at 21, because we have an acquirable gap there.
                    startOffset = Math.min(initialReadGapOffset.gapStartOffset(), startOffset));
}